src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java

```diff
@@ -0,0 +1,258 @@
+    StarlarkOverrideApi rootOverride = overrides.get(module.getName());
+    if (rootOverride != null) {
+      throw errorf("invalid override for the root module found: %s", rootOverride);
+    }
+    ImmutableMap<String, StarlarkOverrideApi> overridesWithRoot =
+        ImmutableMap.<String, StarlarkOverrideApi>builder()
+            .putAll(overrides)
+            .put(module.getName(), LocalPathOverride.create(""))
+            .build();
+
+    return new ModuleFileValue(module, overridesWithRoot);
+  }
+
+  private ModuleFileGlobals execModuleFile(byte[] moduleFile, ModuleKey moduleKey,
+      StarlarkSemantics starlarkSemantics, Environment env)
+      throws ModuleFileFunctionException, InterruptedException {
+    StarlarkFile starlarkFile =
+        StarlarkFile.parse(ParserInput.fromUTF8(moduleFile, moduleKey + "/MODULE.bazel"));
+    if (!starlarkFile.ok()) {
+      Event.replayEventsOn(env.getListener(), starlarkFile.errors());
+      throw errorf("error parsing MODULE.bazel file for %s", moduleKey);
+    }
+
+    ModuleFileGlobals moduleFileGlobals = new ModuleFileGlobals();
+    try (Mutability mu = Mutability.create("module file", moduleKey)) {
+      net.starlark.java.eval.Module predeclaredEnv = getPredeclaredEnv(moduleFileGlobals,
+          starlarkSemantics);
+      Program program = Program.compileFile(starlarkFile, predeclaredEnv);
+      // TODO: check that `program` has no `def`, `if`, etc
+      StarlarkThread thread = new StarlarkThread(mu, starlarkSemantics);
+      thread.setPrintHandler(Event.makeDebugPrintHandler(env.getListener()));
+      Starlark.execFileProgram(program, predeclaredEnv, thread);
+    } catch (SyntaxError.Exception | EvalException e) {
+      throw new ModuleFileFunctionException(e);
+    }
+    return moduleFileGlobals;
+  }
+
+  private static class GetModuleFileResult {
+
+    byte[] moduleFileContents;
+    // `registry` can be null if this module has a non-registry override.
+    @Nullable Registry registry;
+  }
+
+  @Nullable
+  private Optional<GetModuleFileResult> getModuleFile(ModuleKey key, StarlarkOverrideApi override,
+      Environment env) throws ModuleFileFunctionException, InterruptedException {
+    if (override instanceof NonRegistryOverride) {
+      // TODO: The repo name is not necessarily the same as the module name here. Fix later.
+      String repoName = key.getName();
+      RepositoryDirectoryValue repoDir = (RepositoryDirectoryValue) env.getValue(
+          RepositoryDirectoryValue.key(RepositoryName.createFromValidStrippedName(repoName)));
+      if (repoDir == null) {
+        return null;
+      }
+      RootedPath moduleFilePath = RootedPath.toRootedPath(
+          Root.fromPath(repoDir.getPath()), PathFragment.create("MODULE.bazel"));
+      if (env.getValue(FileValue.key(moduleFilePath)) == null) {
+        return Optional.empty();
+      }
+      GetModuleFileResult result = new GetModuleFileResult();
```

Comment:
nit: make `GetModuleFileResult` an immutable object so that you can say `GetModuleFileResult.of(readFile(moduleFilePath.asPath()))`